Update CONTRIBUTING.md (#385)
tenacityteam/tenacity: CONTRIBUTING.md

## Reporting bugs and feature requests

The best place to ask questions about using Tenacity, reporting bugs and
discussing feature requests is the
[tenacity-discuss mailing list](https://lists.sr.ht/~tenacity/tenacity-discuss).
Using a mailing list is as simple as sending an ordinary e-mail. Keep in mind
that all e-mails posted in our mailing list will be archived publicly.
It may be a good idea to look through our archives, because it is possible that
someone else has asked the same question as you have before.

> You can also reach out to us in the room #tenacity at [Libera)](https://web.libera.chat/gamja/?channels=#tenacity)
> if you would like to discuss with us in real time.

Reviewer: Make it a seperate web link? And remove the )
```suggestion
You can also reach out to us in the room #tenacity at Libera.chat [(web)](https://web.libera.chat/gamja/?channels=#tenacity)
if you would like to discuss with us in real time.
```

Author: We could also replace all the links with https://web.libera.chat/#tenacity

## Reporting bugs and feature requests

Our IRC channel, [`#tenacity` on Libera](https://web.libera.chat/#tenacity)
([IRC](ircs://irc.libera.chat:6697/#libera)/
[Matrix](https://matrix.to/#/#tenacity:libera.chat)), is the best place to ask
general questions about Tenacity and talk about feature requests in real time.

You can use [GitHub Issues](https://github.com/tenacityteam/tenacity/issues)

> to report bugs and propose feature requests *only*.

Reviewer 2: I feel like it's worth reiterating this here:

```suggestion
to report bugs and propose feature requests *only*. Support questions will be 
ignored.

You can also reach out to us on [Libera](https://web.libera.chat/#tenacity)
([IRC](ircs://irc.libera.chat:6697/#libera)/
[Matrix](https://matrix.to/#/#tenacity:libera.chat))

> if you would like to discuss with us in real time.

Reviewer 2: I don't really like the way this is laid out.
- "Libera" should be replaced with "irc.libera.chat", no codeblocks, so that people who use IRC but not Libera know how to connect to the server.
- This part of the file is outdated, I think, and as much as I'd like to see people using the mailing lists more, the IRC links need to come first.
- `s/e-mail/email/g`

Author: Libera is the most popular IRC server at the moment, the IRC hyperlink is there, this will basically confuse everyone else apart from the minority that uses IRC but doesn't know of Libera, somehow.

Reviewer: ```suggestion
The best place to ask questions about using Tenacity, discuss feature requests, is the `#tenacity` [Libera](https://web.libera.chat/#tenacity)
([irc.libera.chat](ircs://irc.libera.chat:6697/#libera)/
[Matrix](https://matrix.to/#/#tenacity:libera.chat)) channel, which is real time.

You may also use the
[tenacity-discuss mailing list](https://lists.sr.ht/~tenacity/tenacity-discuss)
for bug reports, feature requests and discussion, but it is not real time.
Using a mailing list is as simple as sending an ordinary email. Keep in mind
that all emails posted in our mailing list will be archived publicly.
It may be a good idea to look through our archives, because it is possible that
someone else has asked the same question as you have before.
```

Reviewer 2: Hmm, linking that way also looks weird. How about this:

```md
The best place to ask questions about using Tenacity and discuss feature
requests is on our IRC channel, [`#tenacity` on
irc.libera.chat](https://web.libera.chat/#tenacity)
([IRC](https://web.libera.chat/#tenacity)/[Matrix](https://matrix.to/#/#tenacity:libera.chat)), 
where we have real-time discussions.     
```

The best place to ask questions about using Tenacity and discuss feature requests is on our IRC channel, [`#tenacity` on irc.libera.chat](https://web.libera.chat/#tenacity) ([IRC](https://web.libera.chat/#tenacity)/[Matrix](https://matrix.to/#/#tenacity:libera.chat)), where we have real-time discussions.

that all emails posted in our mailing list will be archived publicly.
It may be a good idea to look through our archives, because it is possible that

> someone else has asked the same question as you have before.

Reviewer 2: `mailto` link to the mailing list, possibly? Not entirely sure about this.

```suggestion
Using the mailing list is as simple as sending an ordinary email to 
[`~tenacity/tenacity-discuss@lists.sr.ht`](mailto:~tenacity/tenacity-discuss@lists.sr.ht).
Keep in mind that all emails posted in our mailing list will be archived
publicly. It may be a good idea to look through our archives, because it is 
possible that someone else has asked the same question as you have before.

only for those which need further looking into, and only when asked to._

When contributing, please adhere to the following guidelines:

1. Follow proper code formatting guidelines e.g. If the file uses spaces, do not
   change them to tabs.
2. Do not change any variable names unless necessary.
3. Try to keep changes compatible with upstream (Audacity)
   whenever possible.
4. Make concise and accurate commit messages. A commit message should be
   limited to 50 characters and its description limited to 72 characters
   per line, and the message should be able to complete this sentence:

    > This commit will...

    If you need to add any additional context, do so in the commit description.
### Submitting code

### Contributing through sourcehut
#### Sending patches through SourceHut

SourceHut operates on an email-driven workflow, and uses [`git send-email`](https://git-send-email.io) for patch submission. Please send your
patches to
[`~tenacity/tenacity-dev@lists.sr.ht`](mailto:~tenacity/tenacity-dev@lists.sr.ht).
When revising a patch, please use `git commit --amend` and add the `-v2` (increment every revision)
flag.

> ### Contributing through GitHub

Reviewer 3: You `open` pull requests on GH.

_Note: you do not need to open a GitHub issue for every matching contribution,
only for those which need further looking into, and only when asked to._

### Submitting code

#### Sending patches through SourceHut

SourceHut operates on an email-driven workflow, and uses
[`git send-email`](https://git-send-email.io) for patch submission. Please send
your patches to
[`~tenacity/tenacity-dev@lists.sr.ht`](mailto:~tenacity/tenacity-dev@lists.sr.ht).

When revising a patch, please use `git commit --amend` and add the `-v2`
(increment every revision) flag.

#### Making pull requests on GitHub

To contribute code using GitHub, first
[fork this repository](https://github.com/tenacityteam/tenacity)
and make your changes. Please use `git commit --amend` and
`git push -f` for minor changes (only **your** commits).

See [git-rebase.io](https://git-rebase.io) for more details.

### Guidelines for code

When contributing, please adhere to the following guidelines:

> 1. Follow proper code formatting guidelines e.g. If the file uses spaces, do not
>    change them to tabs.
> 2. Do not change any variable names unless necessary.
> 3. Try to keep changes compatible with upstream (Audacity)
>    whenever possible.
> 4. Follow the [commit message guidelines](#commit-messages).

Reviewer: Do we need to have Audacity compatibility this at this point?
```suggestion
1. Follow proper code formatting guidelines e.g. If the file uses spaces, do not
   change them to tabs.
2. Do not change any variable names unless necessary.
43 Follow the [commit message guidelines](#commit-messages).
```

Author: Well, no, but it's like whatever at this point.

Each commit must include a DCO in its description, which looks like this:

```
Signed-off-by: Con Tributor <con.tributor@example.com>
```

You may type this line manually or, if using the command line, simply append
`-s` to your commit.

We ask that all contributors use their real email address, which can be replied
to.

#### Commit messages

The following rules continue to support hosting our code on multiple platforms,
such as both GitHub and SourceHut, without being unnecessarily locked in to
GitHub. Moreover, they are also necessary for complying to the GPL license,
ensuring our independence and just giving credit where it is due.

Apart from including a DCO, as mentioned earlier, the following are also very
important:

* Make concise and accurate commit messages. A commit message should be
   limited to 50 characters and its description limited to 72 characters
   per line, and the message should be able to complete this sentence:

    > This commit will...

    If you need to add any additional context, do so in the commit description.

> * Avoid using dots or past tenses in your commit messages.

Reviewer 4: "Avoid using full stops or past tense in your commit messages."

Author: Yeah, that's a much better way to put it.

  or alias that they have used in Git in the past, as well as their e-mail.

   Example: `Co-authored-by: Jane Doe <jane.doe@example.com>`

* If your commit is complicated and involves multiple changes, use asterisks
  and explain of the changes you made in a few words.

   Example:

   ```
   Prepare Teriyaki Sauce

   * Added Soy Sauce
   * Added Cooking Sake and Sugar
   * Added 1 teaspoon of Mirin
   * Added Dashi Stock
   * Mixed ingredients together
   ```

* If you're using changes that were made by another person, the original
  changes by that said person should generally be signed off and available
  publicly in places such as another pull request on GitHub. Exceptions can
  be made, but do ***not*** sign off a commit for another person without
  their explicit permission.

* Leave an empty line between tags such as `Signed-off-by:` and the rest of
  the commit description.

   Example:

   ```
   Remove references to pumpkin pies

   I get that pumpkin pie is tasty, but this does not have anything to do
   with Tenacity whatsoever.

   Signed-off-by: Pumpkin Hater <pumpkin.hater99@example.com>
   ```

##### GitHub

> * Avoid using emojis or GitHub-specific references (e.g. `:tada:`) to emojis

Reviewer 4: I would say that the emojis are fine in the commit body, but not in the commit subject. Thoughts?

Author: @SFR-git

Reviewer 2: Actual Unicode emoji will cause problems. GitHub ones should be fine, but for simplicity's sake we should disallow them outright.

  in your commits. They may look just fine on GitHub, but they do not anywhere
  else.

##### Maintainers

* When merging pull requests from GitHub, make sure to remove references to
  issues or pull requests that have a numeric format (e.g. `(#1234)` or
  `Resolves #1234`). Including a hyperlink to the said issue or pull request

>   is preferred, because these links will not break outside of GitHub and will

Reviewer 3: What about using `GH #1234`?

Author: It's still harder to find than a copied and pasted link?

  also reduce confusion between patches that refer to issues in the Audacity

>   repository and patches that are meant to be used in Tenacity.

Reviewer 4: Add line (to clarify GitHub behavior so people don't think they made a mistake when GitHub auto-magically replaces the hyperlink): 

- If you use a hyperlink instead of just the `#nnn` format, GitHub will still show the `#nnn` format on the website, but other websites and/or the command line will show the full hyperlink.

such as both GitHub and SourceHut, without being unnecessarily locked in to
GitHub. Moreover, they are also necessary for complying to the GPL license,
ensuring our independence and just giving credit where it is due.

Our stance on these rules is not very strict. Worst case scenario, we may
correct the commit messages for you and inform you about what we had to correct
for future reference. However, if you would like to get seriously involved
with our project and take on responsibilities such as as reviewing and merging
patches, your abidance to the following rules will also be one of the factors
that will be considered.

Apart from including a DCO, as mentioned earlier, the following are also very
important:

* Make concise and accurate commit messages. A commit message should be
   limited to 50 characters and its description limited to 72 characters
   per line, and the message should be able to complete this sentence:

    > This commit will...

    If you need to add any additional context, do so in the commit description.

* Avoid using full stops (e.g. `.`) or past tense in your commit messages.

   - Correct: `Add support for the Commodore 64`

>    - Wrong: `Added support for the Commodore 64.`

Reviewer 2: ```suggestion
   - Incorrect: `Added support for the Commodore 64.`
